### DISCUSSION TITLE
Development Suggestions

### DISCUSSION THREAD

**Discussion by bizehao** [microsoft/WindowsAppSDK#2303] [Q&A]
I wanted to learn the development of winui 3, but I fell into a choice because winui 3 supports both c# and CPP development I'm proficient in both languages. I'd like to ask whether you recommend using c# development or CPP and  Is there a big performance difference between the two

**Comment by castorix** [comment]
C++ is C++/WinRT and C# is a lot simpler for WinUI (while I use C/C++since 80's...)
Also most of MS samples are in C# (like [XAML Controls Gallery](https://github.com/microsoft/Xaml-Controls-Gallery) (UWP source for WinUI 3 Controls Gallery))

**Reply by bizehao** [reply]
Is there a big difference in performance between the two?

**Reply by castorix** [reply]
> Is there a big difference in performance between the two?

I did not notice any difference (I have even tested in VB, with some animations to test, and it works fine)
But I did not test with big data (like in ListViews for example)

**Comment by ghost** [comment]
> Is there a big difference in performance between the two?

**TL;DR :** Sorry. Don't know. 
Long Story short,, if performance is not a huge concern, I suggest doing WinUI 3 devs(only UI part) in C#/WinRT.

---
.xaml(winui 3)  files have to go through 
👉 .xaml ->.idl -> .winmd -> .h/.cpp (for C++) || .cs (for C#) ||  .rs/.py (for Rust/Python , whenever these two projections support  compiling .winmd files).

👉👉 essentially  it(winui 3's performance) all boils down to "language projections" performance.

now,  offcourse C++/WinRT is faster than C#/WinRT (without NativeAOT). so C++/WinRT/WinUI 3 app will be faster naturally.

can't provide concreate C++/WinRT vs C#/WinRT(without NativeAOT) performance comparison numbers.

---
* until there's any direct way is provided to generate .winmd files from .xaml files. so you will have to write .idl files manually for each languages other than C#. 

[C# does not require IDL for it's reflection and maybe because it XAML was born in C# ?. I'm fairly confident that whenver C#/WinRT starts to support NativeAOT , dealing with .idl files will be required and/or tools for directly generating .winmd files will be provided.]

* VS support for (midl3) .idl files  is nonexistent. VS Code has a fine extension.
* manually copying and pasting .h/.cpp files into your solution (generated from .winmd files by cppwinrt.exe tool) will be pain. because VS doesn't have deep support for C++/WinRT. 

why C++/WinRT support is so poor in VS ? it's because of C++/CX or C#/WinRT Politics, nothing else.
